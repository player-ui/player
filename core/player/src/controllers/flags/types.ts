import { Severity } from "../../logger";

/** Configuration for runtime Player behavior */
export interface PlayerFlags {
  /** What log level duplicate ID errors during view resolution should be raised as */
  duplicateIDLogLevel: Severity;

  /** Log level for when there are cache conflicts in view resolution (usually related to duplicate IDs) */
  cacheConflictLogLevel: Severity;
}

export const DefaultFlags: PlayerFlags = {
  duplicateIDLogLevel: "error",
  cacheConflictLogLevel: "info",
};

export type Flag = keyof PlayerFlags;
