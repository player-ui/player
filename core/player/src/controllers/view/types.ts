import type { Asset } from "@player-ui/types";
import type { Registry } from "@player-ui/partial-match-registry";

import type { Resolve, Node } from "../../view";
import type { Store } from "./store";

/** Transform function that is ran on the Asset before it's resolved */
export type BeforeTransformFunction<AuthoredAsset extends Asset = Asset> = (
  asset: Node.Asset<AuthoredAsset> | Node.View<AuthoredAsset>,
  options: Resolve.NodeResolveOptions,
  store: Store,
) => Node.Node;

/** Transform function that is ran on the Asset after it's resolved */
export type TransformFunction<
  AuthoredAsset extends Asset = Asset,
  TransformedAsset extends Asset = AuthoredAsset,
> = (
  asset: AuthoredAsset,
  options: Resolve.NodeResolveOptions,
  store: Store,
) => TransformedAsset;

export interface TransformFunctions {
  /** A function that is executed as an AST -> AST transform before resolving the node to a value */
  beforeResolve?: BeforeTransformFunction<any>;

  /** A function to resolve an AST to a value */
  resolve?: TransformFunction<any>;
}

export type TransformRegistry = Registry<TransformFunctions>;
