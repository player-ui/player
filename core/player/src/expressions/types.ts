import type { DataModelWithParser } from '../data';
import type { Logger } from '../logger';

export type ExpressionLiteralType =
  | string
  | number
  | boolean
  | undefined
  | null;
export type ExpressionType =
  | object
  | ExpressionLiteralType
  | Array<ExpressionLiteralType>
  | ExpressionNode;

export interface OperatorProcessingOptions {
  /**
   * When set to a falsy value, the arguments passed to the handler will be raw AST Nodes
   * This enables lazy evaluation of arguments
   */
  resolveParams: boolean;
}

export type BinaryOperatorBasic = (left: any, right: any) => unknown;
export type BinaryOperatorAdvanced = OperatorProcessingOptions &
  ((ctx: ExpressionContext, left: any, right: any) => unknown);

export type BinaryOperator = BinaryOperatorAdvanced | BinaryOperatorBasic;

export type UnaryOperator =
  | ((arg: any) => unknown)
  | (((ctx: ExpressionContext, arg: any) => unknown) &
      OperatorProcessingOptions);

export interface ExpressionContext {
  /** A means of executing an expression */
  evaluate: (expr: ExpressionType) => unknown;

  /** The data model that expression handlers can use when fetching data */
  model: DataModelWithParser;

  /** A logger to use */
  logger?: Logger;
}

export type ExpressionHandler<
  T extends readonly unknown[] = unknown[],
  R = void
> = ((context: ExpressionContext, ...args: T) => R) &
  Partial<OperatorProcessingOptions>;

export const ExpNodeOpaqueIdentifier = Symbol('Expression Node ID');

/** Checks if the input is an already processed Expression node */
export function isExpressionNode(x: any): x is ExpressionNode {
  return typeof x === 'object' && x.__id === ExpNodeOpaqueIdentifier;
}

export interface BaseNode<T> {
  /** The thing to discriminate the AST type on */
  type: T;

  /** How to tell this apart from other objects */
  __id: typeof ExpNodeOpaqueIdentifier;
}

/** A helper interface for nodes that container left and right children */
export interface DirectionalNode {
  /** The left node. Often for the left hand side of an expression */
  left: ExpressionNode;

  /** The right child. Often for the right hand side of an expression */
  right: ExpressionNode;
}

export interface LiteralNode extends BaseNode<'Literal'> {
  /** A node that holds a literal value */
  value: string | number;

  /** The unprocessed value */
  raw?: any;
}

export interface BinaryNode
  extends BaseNode<'BinaryExpression'>,
    DirectionalNode {
  /** The operation to perform on the nodes */
  operator: string;
}

export interface LogicalNode extends BaseNode<'LogicalExpression'> {
  /** The left hand side of the equation */
  left: any;

  /** The right hand side of the equation */
  right: any;

  /** The logical operation to perform on the nodes */
  operator: string;
}

export interface UnaryNode extends BaseNode<'UnaryExpression'> {
  /** The operation to perform on the node */
  operator: string;

  /** The single argument that the operation should be performed on */
  argument: any;
}

export type ThisNode = BaseNode<'ThisExpression'>;

export interface ModelRefNode extends BaseNode<'ModelRef'> {
  /** The binding that the model reference points to */
  ref: string;
}

export interface ObjectNode extends BaseNode<'Object'> {
  /**  */
  attributes: Array<{
    /** The property name of the object */
    key: any;

    /** the associated value */
    value: any;
  }>;
}

export interface MemberExpressionNode extends BaseNode<'MemberExpression'> {
  /** The object to be introspected */
  object: ExpressionNode;

  /** If the property uses . or open-bracket */
  computed: boolean;

  /** The property to access on the object */
  property: ExpressionNode;
}

export interface ConditionalExpressionNode
  extends BaseNode<'ConditionalExpression'> {
  /** The test for the ternary */
  test: ExpressionNode;

  /** The truthy case for the ternary */
  consequent: ExpressionNode;

  /** The falsy case for the ternary */
  alternate: ExpressionNode;
}

export interface CompoundNode extends BaseNode<'Compound'> {
  /** The contents of the compound expression */
  body: ExpressionNode[];
}

export interface CallExpressionNode extends BaseNode<'CallExpression'> {
  /** The arguments to the function */
  args: any[];

  /** The function name */
  callTarget: IdentifierNode;
}

export interface ArrayExpressionNode extends BaseNode<'ArrayExpression'> {
  /** The items in an array */
  elements: any[];
}

export interface IdentifierNode extends BaseNode<'Identifier'> {
  /** The variable name */
  name: string;
}

export type AssignmentNode = BaseNode<'Assignment'> & DirectionalNode;

export interface ModificationNode
  extends BaseNode<'Modification'>,
    DirectionalNode {
  /** The operator for the modification */
  operator: string;
}

export type ExpressionNode =
  | LiteralNode
  | BinaryNode
  | LogicalNode
  | UnaryNode
  | ThisNode
  | ModelRefNode
  | MemberExpressionNode
  | ConditionalExpressionNode
  | CompoundNode
  | CallExpressionNode
  | ArrayExpressionNode
  | IdentifierNode
  | AssignmentNode
  | ModificationNode
  | ObjectNode;

export type ExpressionNodeType = ExpressionNode['type'];
