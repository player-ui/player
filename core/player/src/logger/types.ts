export type LogFn = (...args: Array<any>) => void;

export const severities = ['trace', 'debug', 'info', 'warn', 'error'] as const;
export type Severity = typeof severities[number];
export type Logger = Record<Severity, LogFn>;
export type LoggerProvider = () => Logger | undefined;
