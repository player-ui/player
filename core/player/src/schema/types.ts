import type { Formatting } from '@player-ui/types';

export type FormatOptions = Omit<Formatting.Reference, 'type'>;

/**
 * The return types for the schema don't include options.
 * These are already sent to the generic formatter function for that type
 */
export type FormatFunction<From, To = From, Options = unknown> = (
  val: From,
  options?: Options
) => To | undefined;

export type FormatHandler<From, To = From> = (val: From) => To;

export interface FormatDefinition<DataModelType, UserDisplayType> {
  /**
   * A function to format data (from the data-model to the user).
   * Defaults to the identify function
   */
  format: FormatHandler<DataModelType, UserDisplayType>;

  /**
   * A function to invert the formatting (from the user to the data-model)
   * Defaults to the identify function.
   */
  deformat: FormatHandler<UserDisplayType, DataModelType>;
}

export interface FormatType<
  DataModelType,
  UserDisplayType = DataModelType,
  Options = undefined
> {
  /**
   * The name of the formatter.
   * This corresponds to the 'type' format property when creating a DataType
   */
  name: string;

  /**
   * An optional function to format data for display to the user.
   * This goes from dataModel -> UI display
   */
  format?: FormatFunction<
    DataModelType | UserDisplayType,
    UserDisplayType,
    Options
  >;

  /**
   * An optional function for undo the action of a format function for storage.
   * This goes from UI -> dataModel
   */
  deformat?: FormatFunction<
    UserDisplayType | DataModelType,
    DataModelType,
    Options
  >;
}
