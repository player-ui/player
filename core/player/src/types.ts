import type { Flow, FlowResult } from '@player-ui/types';
import type { BindingParser, BindingLike } from './binding';
import type { SchemaController } from './schema';
import type { ExpressionEvaluator } from './expressions';
import type { Logger } from './logger';
import type {
  ViewController,
  DataController,
  ValidationController,
  FlowController,
} from './controllers';
import type { ReadOnlyDataController } from './controllers/data/utils';

/** The status for a flow's execution state */
export type PlayerFlowStatus =
  | 'not-started'
  | 'in-progress'
  | 'completed'
  | 'error';

/** Common interface for the state of Player's flow execution */
export interface BaseFlowState<T extends PlayerFlowStatus> {
  /** A unique reference for the life-cycle of a flow */
  ref: symbol;

  /** The status of the given flow */
  status: T;
}

/** The beginning state of Player, before it's seen a flow  */
export type NotStartedState = BaseFlowState<'not-started'>;

export const NOT_STARTED_STATE: NotStartedState = {
  ref: Symbol('not-started'),
  status: 'not-started',
};

/** Shared properties for a flow in any state of execution (in-progress, completed successfully, or errored out) */
export interface PlayerFlowExecutionData {
  /** The currently executing flow */
  flow: Flow;
}

export interface ControllerState {
  /** The manager for data for a flow */
  data: DataController;

  /** The view manager for a flow */
  view: ViewController;

  /** The schema manager for a flow */
  schema: SchemaController;

  /** The validation manager for a flow */
  validation: ValidationController;

  /** The expression evaluator for a flow */
  expression: ExpressionEvaluator;

  /** The manager for parsing and resolving bindings */
  binding: BindingParser;

  /** the manager for the flow state machine */
  flow: FlowController;
}

/** A flow is currently executing */
export type InProgressState = BaseFlowState<'in-progress'> &
  PlayerFlowExecutionData & {
    /** A promise that resolves when the flow is completed */
    flowResult: Promise<FlowResult>;

    /** The underlying state controllers for the current flow */
    controllers: ControllerState;

    /** Allow other platforms to abort the current flow with an error  */
    fail: (error: Error) => void;

    /**
     * The Logger for the current player instance
     */
    logger: Logger;
  };

/** The flow completed properly */
export type CompletedState = BaseFlowState<'completed'> &
  PlayerFlowExecutionData &
  FlowResult & {
    /** Readonly Player controllers to provide Player functionality after the flow has ended */
    controllers: {
      /** A read only instance of the Data Controller */
      data: ReadOnlyDataController;
    };
  };

/** The flow finished but not successfully */
export type ErrorState = BaseFlowState<'error'> & {
  /** The currently executing flow */
  flow: Flow;

  /** The error associated with the failed flow */
  error: Error;
};

/** Any Player state  */
export type PlayerFlowState =
  | NotStartedState
  | InProgressState
  | CompletedState
  | ErrorState;

// Model

export type RawSetType = [BindingLike, any];
export type RawSetTransaction = Record<string, any> | RawSetType[];
