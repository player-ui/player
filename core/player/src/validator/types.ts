import type { Validation } from '@player-ui/types';

import type { BindingInstance, BindingFactory } from '../binding';
import type { DataModelWithParser } from '../data';
import type { ExpressionEvaluatorFunction } from '../expressions';
import type { Logger } from '../logger';
import type { ConstantsProvider } from '../controllers';

interface BaseValidationResponse<T = Validation.Severity> {
  /** The validation message to show to the user */
  message: string;

  /** List of parameters associated with a validation. These can be replaced into a templatized message string. */
  parameters?: Record<string, any>;

  /** How serious is this violation */
  severity: T;

  /** Where this validation should be displayed */
  displayTarget?: Validation.DisplayTarget;

  /** The blocking state of this validation */
  blocking?: boolean | 'once';
}

export interface WarningValidationResponse
  extends BaseValidationResponse<'warning'> {
  /** Warning validations can be dismissed without correcting the error */
  dismiss?: () => void;
}

export type ErrorValidationResponse = BaseValidationResponse<'error'>;

export type ValidationResponse =
  | ErrorValidationResponse
  | WarningValidationResponse;

type RequiredValidationKeys = 'severity' | 'trigger';

export type ValidationObject = Validation.Reference &
  Required<Pick<Validation.Reference, RequiredValidationKeys>>;

export interface ValidationProvider {
  getValidationsForBinding?(
    binding: BindingInstance
  ): Array<ValidationObject> | undefined;

  getValidationsForView?(): Array<ValidationObject> | undefined;
}

export interface ValidatorContext {
  /** The data to set or get data from */
  model: DataModelWithParser;

  /** A means of parsing a binding */
  parseBinding: BindingFactory;

  /** Execute the expression and return it's result */
  evaluate: ExpressionEvaluatorFunction;

  /** Logger instance to use */
  logger: Logger;

  /** The validation object that triggered this function */
  validation: ValidationObject;

  /** The constants for messages */
  constants: ConstantsProvider;
}

export type ValidatorFunction<Options = unknown> = (
  context: ValidatorContext,
  value: any,
  options?: Options
) => Omit<BaseValidationResponse, 'severity'> | undefined;
