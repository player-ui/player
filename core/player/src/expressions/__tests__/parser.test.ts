import { parseExpression } from '../parser';
import { ExpNodeOpaqueIdentifier } from '../types';

test('the happy stuff', () => {
  expect(parseExpression('foo')).toMatchSnapshot();
  expect(parseExpression('foo === bar')).toMatchSnapshot();
  expect(parseExpression('foo || bar')).toMatchSnapshot();
  expect(parseExpression('foo(bar)')).toMatchSnapshot();
  expect(parseExpression('{{foo}} = "bar"')).toMatchSnapshot();
  expect(parseExpression('{{foo}} ? "bar" : "baz"')).toMatchSnapshot();
  expect(parseExpression('foo[bar]')).toMatchSnapshot();
  expect(
    parseExpression('{{foo}} == "string\nwith\tbreaks"')
  ).toMatchSnapshot();
  expect(parseExpression('foo = [1, 2, 3]')).toMatchSnapshot();
});

test('nested binary op location', () => {
  expect(parseExpression('foo === bar === baz')).toStrictEqual({
    __id: ExpNodeOpaqueIdentifier,
    type: 'BinaryExpression',
    operator: '===',
    left: {
      __id: ExpNodeOpaqueIdentifier,
      type: 'BinaryExpression',
      operator: '===',
      left: {
        __id: ExpNodeOpaqueIdentifier,
        type: 'Identifier',
        name: 'foo',
        location: {
          start: { character: 0 },
          end: { character: 3 },
        },
      },
      right: {
        __id: ExpNodeOpaqueIdentifier,
        type: 'Identifier',
        name: 'bar',
        location: {
          start: { character: 8 },
          end: { character: 11 },
        },
      },
      location: {
        start: { character: 0 },
        end: { character: 11 },
      },
    },
    right: {
      __id: ExpNodeOpaqueIdentifier,
      type: 'Identifier',
      name: 'baz',
      location: {
        start: { character: 16 },
        end: { character: 19 },
      },
    },
    location: {
      start: { character: 0 },
      end: { character: 19 },
    },
  });
});

test('the bad stuff', () => {
  expect(() => parseExpression('{{foo')).toThrowError();
  expect(() => parseExpression('{{foo}')).toThrowError();
  expect(() => parseExpression('foo["bar"')).toThrowError();
  expect(() => parseExpression('foo["bar')).toThrowError();
  expect(() => parseExpression('foo]')).toThrowError();
  expect(() => parseExpression('foo = (bar')).toThrowError();
  expect(() => parseExpression('foo(')).toThrowError();
});
describe('expression parser', () => {
  test('objects- in parser', () => {
    expect(parseExpression('{"foo": "value"}')).toMatchSnapshot();
    expect(parseExpression('{"foo": {{some.binding}}}')).toMatchSnapshot();
    expect(parseExpression('{"foo": 1 + 2}')).toMatchSnapshot();
    expect(parseExpression('{"foo": "value", "bar": "baz"}')).toMatchSnapshot();
    expect(
      parseExpression('{"foo": "value", "bar": { "baz":  "foo" }}')
    ).toMatchSnapshot();
    expect(
      parseExpression('publish("test", {"key": "value"})')
    ).toMatchSnapshot();
    // no closing brace
    expect(() => parseExpression('{"foo": "value"')).toThrowError();
    // no key
    expect(() => parseExpression('{: "value"}')).toThrowError();
    // no value
    expect(() => parseExpression('{"key": }')).toThrowError();
    // no colon
    expect(() => parseExpression('{"key" "value" }')).toThrowError();
    // no comma
    expect(() =>
      parseExpression('{"key": "value" "key2": "value" }')
    ).toThrowError();
  });
});

describe('graceful parsing', () => {
  it('returns and sets the error for invalid expression', () => {
    const parsed = parseExpression('{{foo}} = {{bar}', {
      strict: false,
    });

    expect(parsed).toBeTruthy();
    expect(parsed.error).toBeTruthy();
    expect(parsed.error?.message).toBe(
      'Unclosed brace after "bar}" at character 16'
    );
  });
});
