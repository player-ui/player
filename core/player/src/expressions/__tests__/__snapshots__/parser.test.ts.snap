// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`expression parser > objects- in parser 1`] = `
{
  "__id": Symbol(Expression Node ID),
  "attributes": [
    {
      "key": {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 6,
          },
          "start": {
            "character": 2,
          },
        },
        "raw": ""foo"",
        "type": "Literal",
        "value": "foo",
      },
      "value": {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 15,
          },
          "start": {
            "character": 9,
          },
        },
        "raw": ""value"",
        "type": "Literal",
        "value": "value",
      },
    },
  ],
  "location": {
    "end": {
      "character": 16,
    },
    "start": {
      "character": 0,
    },
  },
  "type": "Object",
}
`;

exports[`expression parser > objects- in parser 2`] = `
{
  "__id": Symbol(Expression Node ID),
  "attributes": [
    {
      "key": {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 6,
          },
          "start": {
            "character": 2,
          },
        },
        "raw": ""foo"",
        "type": "Literal",
        "value": "foo",
      },
      "value": {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 24,
          },
          "start": {
            "character": 8,
          },
        },
        "ref": "some.binding",
        "type": "ModelRef",
      },
    },
  ],
  "location": {
    "end": {
      "character": 25,
    },
    "start": {
      "character": 0,
    },
  },
  "type": "Object",
}
`;

exports[`expression parser > objects- in parser 3`] = `
{
  "__id": Symbol(Expression Node ID),
  "attributes": [
    {
      "key": {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 6,
          },
          "start": {
            "character": 2,
          },
        },
        "raw": ""foo"",
        "type": "Literal",
        "value": "foo",
      },
      "value": {
        "__id": Symbol(Expression Node ID),
        "left": {
          "__id": Symbol(Expression Node ID),
          "location": {
            "end": {
              "character": 9,
            },
            "start": {
              "character": 8,
            },
          },
          "raw": "1",
          "type": "Literal",
          "value": 1,
        },
        "location": {
          "end": {
            "character": 13,
          },
          "start": {
            "character": 8,
          },
        },
        "operator": "+",
        "right": {
          "__id": Symbol(Expression Node ID),
          "location": {
            "end": {
              "character": 13,
            },
            "start": {
              "character": 12,
            },
          },
          "raw": "2",
          "type": "Literal",
          "value": 2,
        },
        "type": "BinaryExpression",
      },
    },
  ],
  "location": {
    "end": {
      "character": 14,
    },
    "start": {
      "character": 0,
    },
  },
  "type": "Object",
}
`;

exports[`expression parser > objects- in parser 4`] = `
{
  "__id": Symbol(Expression Node ID),
  "attributes": [
    {
      "key": {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 6,
          },
          "start": {
            "character": 2,
          },
        },
        "raw": ""foo"",
        "type": "Literal",
        "value": "foo",
      },
      "value": {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 15,
          },
          "start": {
            "character": 9,
          },
        },
        "raw": ""value"",
        "type": "Literal",
        "value": "value",
      },
    },
    {
      "key": {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 22,
          },
          "start": {
            "character": 18,
          },
        },
        "raw": ""bar"",
        "type": "Literal",
        "value": "bar",
      },
      "value": {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 29,
          },
          "start": {
            "character": 25,
          },
        },
        "raw": ""baz"",
        "type": "Literal",
        "value": "baz",
      },
    },
  ],
  "location": {
    "end": {
      "character": 30,
    },
    "start": {
      "character": 0,
    },
  },
  "type": "Object",
}
`;

exports[`expression parser > objects- in parser 5`] = `
{
  "__id": Symbol(Expression Node ID),
  "attributes": [
    {
      "key": {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 6,
          },
          "start": {
            "character": 2,
          },
        },
        "raw": ""foo"",
        "type": "Literal",
        "value": "foo",
      },
      "value": {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 15,
          },
          "start": {
            "character": 9,
          },
        },
        "raw": ""value"",
        "type": "Literal",
        "value": "value",
      },
    },
    {
      "key": {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 22,
          },
          "start": {
            "character": 18,
          },
        },
        "raw": ""bar"",
        "type": "Literal",
        "value": "bar",
      },
      "value": {
        "__id": Symbol(Expression Node ID),
        "attributes": [
          {
            "key": {
              "__id": Symbol(Expression Node ID),
              "location": {
                "end": {
                  "character": 31,
                },
                "start": {
                  "character": 27,
                },
              },
              "raw": ""baz"",
              "type": "Literal",
              "value": "baz",
            },
            "value": {
              "__id": Symbol(Expression Node ID),
              "location": {
                "end": {
                  "character": 39,
                },
                "start": {
                  "character": 35,
                },
              },
              "raw": ""foo"",
              "type": "Literal",
              "value": "foo",
            },
          },
        ],
        "location": {
          "end": {
            "character": 41,
          },
          "start": {
            "character": 24,
          },
        },
        "type": "Object",
      },
    },
  ],
  "location": {
    "end": {
      "character": 42,
    },
    "start": {
      "character": 0,
    },
  },
  "type": "Object",
}
`;

exports[`expression parser > objects- in parser 6`] = `
{
  "__id": Symbol(Expression Node ID),
  "args": [
    {
      "__id": Symbol(Expression Node ID),
      "location": {
        "end": {
          "character": 14,
        },
        "start": {
          "character": 9,
        },
      },
      "raw": ""test"",
      "type": "Literal",
      "value": "test",
    },
    {
      "__id": Symbol(Expression Node ID),
      "attributes": [
        {
          "key": {
            "__id": Symbol(Expression Node ID),
            "location": {
              "end": {
                "character": 22,
              },
              "start": {
                "character": 18,
              },
            },
            "raw": ""key"",
            "type": "Literal",
            "value": "key",
          },
          "value": {
            "__id": Symbol(Expression Node ID),
            "location": {
              "end": {
                "character": 31,
              },
              "start": {
                "character": 25,
              },
            },
            "raw": ""value"",
            "type": "Literal",
            "value": "value",
          },
        },
      ],
      "location": {
        "end": {
          "character": 32,
        },
        "start": {
          "character": 16,
        },
      },
      "type": "Object",
    },
  ],
  "callTarget": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 7,
      },
      "start": {
        "character": 0,
      },
    },
    "name": "publish",
    "type": "Identifier",
  },
  "location": {
    "end": {
      "character": 33,
    },
    "start": {
      "character": 7,
    },
  },
  "type": "CallExpression",
}
`;

exports[`the happy stuff 1`] = `
{
  "__id": Symbol(Expression Node ID),
  "location": {
    "end": {
      "character": 3,
    },
    "start": {
      "character": 0,
    },
  },
  "name": "foo",
  "type": "Identifier",
}
`;

exports[`the happy stuff 2`] = `
{
  "__id": Symbol(Expression Node ID),
  "left": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 3,
      },
      "start": {
        "character": 0,
      },
    },
    "name": "foo",
    "type": "Identifier",
  },
  "location": {
    "end": {
      "character": 11,
    },
    "start": {
      "character": 0,
    },
  },
  "operator": "===",
  "right": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 11,
      },
      "start": {
        "character": 8,
      },
    },
    "name": "bar",
    "type": "Identifier",
  },
  "type": "BinaryExpression",
}
`;

exports[`the happy stuff 3`] = `
{
  "__id": Symbol(Expression Node ID),
  "left": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 3,
      },
      "start": {
        "character": 0,
      },
    },
    "name": "foo",
    "type": "Identifier",
  },
  "location": {
    "end": {
      "character": 10,
    },
    "start": {
      "character": 0,
    },
  },
  "operator": "||",
  "right": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 10,
      },
      "start": {
        "character": 7,
      },
    },
    "name": "bar",
    "type": "Identifier",
  },
  "type": "LogicalExpression",
}
`;

exports[`the happy stuff 4`] = `
{
  "__id": Symbol(Expression Node ID),
  "args": [
    {
      "__id": Symbol(Expression Node ID),
      "location": {
        "end": {
          "character": 7,
        },
        "start": {
          "character": 4,
        },
      },
      "name": "bar",
      "type": "Identifier",
    },
  ],
  "callTarget": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 3,
      },
      "start": {
        "character": 0,
      },
    },
    "name": "foo",
    "type": "Identifier",
  },
  "location": {
    "end": {
      "character": 8,
    },
    "start": {
      "character": 3,
    },
  },
  "type": "CallExpression",
}
`;

exports[`the happy stuff 5`] = `
{
  "__id": Symbol(Expression Node ID),
  "left": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 7,
      },
      "start": {
        "character": 0,
      },
    },
    "ref": "foo",
    "type": "ModelRef",
  },
  "location": {
    "end": {
      "character": 15,
    },
    "start": {
      "character": 0,
    },
  },
  "operator": "=",
  "right": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 15,
      },
      "start": {
        "character": 11,
      },
    },
    "raw": ""bar"",
    "type": "Literal",
    "value": "bar",
  },
  "type": "Assignment",
}
`;

exports[`the happy stuff 6`] = `
{
  "__id": Symbol(Expression Node ID),
  "alternate": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 23,
      },
      "start": {
        "character": 19,
      },
    },
    "raw": ""baz"",
    "type": "Literal",
    "value": "baz",
  },
  "consequent": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 15,
      },
      "start": {
        "character": 11,
      },
    },
    "raw": ""bar"",
    "type": "Literal",
    "value": "bar",
  },
  "location": {
    "end": {
      "character": 23,
    },
    "start": {
      "character": 8,
    },
  },
  "test": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 7,
      },
      "start": {
        "character": 0,
      },
    },
    "ref": "foo",
    "type": "ModelRef",
  },
  "type": "ConditionalExpression",
}
`;

exports[`the happy stuff 7`] = `
{
  "__id": Symbol(Expression Node ID),
  "computed": true,
  "location": {
    "end": {
      "character": 7,
    },
    "start": {
      "character": 3,
    },
  },
  "object": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 3,
      },
      "start": {
        "character": 0,
      },
    },
    "name": "foo",
    "type": "Identifier",
  },
  "property": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 7,
      },
      "start": {
        "character": 4,
      },
    },
    "name": "bar",
    "type": "Identifier",
  },
  "type": "MemberExpression",
}
`;

exports[`the happy stuff 8`] = `
{
  "__id": Symbol(Expression Node ID),
  "left": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 7,
      },
      "start": {
        "character": 0,
      },
    },
    "ref": "foo",
    "type": "ModelRef",
  },
  "location": {
    "end": {
      "character": 31,
    },
    "start": {
      "character": 0,
    },
  },
  "operator": "==",
  "right": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 31,
      },
      "start": {
        "character": 12,
      },
    },
    "raw": ""string
with	breaks"",
    "type": "Literal",
    "value": "string
with	breaks",
  },
  "type": "BinaryExpression",
}
`;

exports[`the happy stuff 9`] = `
{
  "__id": Symbol(Expression Node ID),
  "left": {
    "__id": Symbol(Expression Node ID),
    "location": {
      "end": {
        "character": 3,
      },
      "start": {
        "character": 0,
      },
    },
    "name": "foo",
    "type": "Identifier",
  },
  "location": {
    "end": {
      "character": 15,
    },
    "start": {
      "character": 0,
    },
  },
  "operator": "=",
  "right": {
    "__id": Symbol(Expression Node ID),
    "elements": [
      {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 8,
          },
          "start": {
            "character": 7,
          },
        },
        "raw": "1",
        "type": "Literal",
        "value": 1,
      },
      {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 11,
          },
          "start": {
            "character": 10,
          },
        },
        "raw": "2",
        "type": "Literal",
        "value": 2,
      },
      {
        "__id": Symbol(Expression Node ID),
        "location": {
          "end": {
            "character": 14,
          },
          "start": {
            "character": 13,
          },
        },
        "raw": "3",
        "type": "Literal",
        "value": 3,
      },
    ],
    "location": {
      "end": {
        "character": 15,
      },
      "start": {
        "character": 6,
      },
    },
    "type": "ArrayExpression",
  },
  "type": "Assignment",
}
`;
